src/query_plan/operation.rs

```diff
@@ -2776,6 +2965,49 @@ impl NormalizedFieldSelection {
     pub(crate) fn has_defer(&self) -> bool {
         self.field.has_defer() || self.selection_set.as_ref().is_some_and(|s| s.has_defer())
     }
+
+    fn can_add_to(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        schema: &ValidFederationSchema,
+    ) -> Result<bool, FederationError> {
+        match (parent_type, &self.field.data().field_position) {
+            (
+                CompositeTypeDefinitionPosition::Object(ty_a),
+                FieldDefinitionPosition::Object(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            (
+                CompositeTypeDefinitionPosition::Interface(ty_a),
+                FieldDefinitionPosition::Interface(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            (
+                CompositeTypeDefinitionPosition::Union(ty_a),
+                FieldDefinitionPosition::Union(ty_b),
+            ) if ty_a.type_name == ty_b.type_name => return Ok(true),
+            _ => {}
+        }
+
+        let Some(ty) = self.field.type_if_added_to(parent_type, schema)? else {
+            return Ok(false);
+        };
+
+        if let Some(set) = &self.selection_set {
+            if set.type_position != ty {
```

Comment:
do we need to unwrap the `ty`? i.e. the `baseType` from JS side of things - I believe there is already a corresponding method for this in `apollo-rs`

Reply:
The JS code does call `baseType`, but I don't think it is needed here. The equivalent is `Type::inner_named_type`, which return the `NamedType` (i.e. a `Name`) san any markers of being nullable or in a list. Here, `ty` is a composition type position, so that extra data should be present.